src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -4061,24 +4061,24 @@ void MacroAssembler::zero_dcache_blocks(Register base, Register cnt, Register tm
   bge(cnt, tmp1, loop);
 }
 
-#define FCVT_SAFE(FLOATCVT, FLOATEQ)                                                             \
-void MacroAssembler:: FLOATCVT##_safe(Register dst, FloatRegister src, Register tmp) {           \
-  Label L_Okay;                                                                                  \
-  fscsr(zr);                                                                                     \
-  FLOATCVT(dst, src);                                                                            \
-  frcsr(tmp);                                                                                    \
-  andi(tmp, tmp, 0x1E);                                                                          \
-  beqz(tmp, L_Okay);                                                                             \
-  FLOATEQ(tmp, src, src);                                                                        \
-  bnez(tmp, L_Okay);                                                                             \
-  mv(dst, zr);                                                                                   \
-  bind(L_Okay);                                                                                  \
-}
-
-FCVT_SAFE(fcvt_w_s, feq_s)
-FCVT_SAFE(fcvt_l_s, feq_s)
-FCVT_SAFE(fcvt_w_d, feq_d)
-FCVT_SAFE(fcvt_l_d, feq_d)
+#define FCVT_SAFE(FLOATCVT, FLOATSIG)                                                     \
+void MacroAssembler::FLOATCVT##_safe(Register dst, FloatRegister src, Register tmp) {     \
+  Label do_convert, done;                                                                 \
+  fclass_##FLOATSIG(tmp, src);                                                            \
+  /* check if src is NaN */                                                               \
+  andi(tmp, tmp, 0b1100000000);                                                           \
+  beqz(tmp, do_convert);                                                                  \
+  mv(dst, zr);                                                                            \
+  j(done);                                                                                \
+  bind(do_convert);                                                                       \
+  FLOATCVT(dst, src);                                                                     \
+  bind(done);                                                                             \
```

Comment:
what about reducing the branching?

e.g.

mv (dst, zr); //pretty cheap anyway
fclass(..);
andi(tmp, tmp, 0b1100000000);
bnez(tmp, done);
  FLOATCVT(dst, src);                                                                    
bind(done);

Reply:
After applying this results look better:

```
Benchmark                     (size)   Mode  Cnt    Score   Error   Units
FloatConversion.doubleToInt     2048  thrpt   15  286.038 ± 1.472  ops/ms
FloatConversion.doubleToLong    2048  thrpt   15  289.585 ± 1.501  ops/ms
FloatConversion.floatToInt      2048  thrpt   15  294.313 ± 1.263  ops/ms
FloatConversion.floatToLong     2048  thrpt   15  273.749 ± 2.261  ops/ms
```

Stable.